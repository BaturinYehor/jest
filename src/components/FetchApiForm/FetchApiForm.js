import React from "react";

const API = 'https://961826c1-030a-4d64-9117-ed8352256d12.mock.pstmn.io/info';

class Main extends React.Component {

    constructor(props) {
        super(props);

        this.state = {
            response: [],
            error: []
        }
    };

    getInformation = (e) => {
        e.preventDefault();
        this.fetchApi(API);
    };

    fetchApi(api) {
        fetch(api)
            .then(response => {
                if(response.ok) {
                    response.json()
                        .then((data) => {
                        if (data.info) {
                            this.setState({response: data.info});//information fetched from api
                        } else {
                            this.setState({error: data.error.message})//can't find api to fetch
                        }
                    })
                } else {
                    throw new Error('Error while fetching api');
                }
            })
            .catch((error) => {
                this.setState({error: error})//error while fetching information from api
            });
    }

    render() {
        return (
            <div>
                <button onClick={this.getInformation}>
                    What is jest?
                </button>
                <div className="get-api-form">
                    <div className="text">
                        {this.state.response}
                        <div className="error">
                            {this.state.error}
                        </div>
                    </div>
                </div>
            </div>
        )
    };
}

export default Main