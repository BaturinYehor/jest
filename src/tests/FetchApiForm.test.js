import {mount} from "enzyme/build";
import FetchApiForm from "../components/FetchApiForm/FetchApiForm";
import React from "react";
import renderer from "react-test-renderer";

const URL = "https://someurl";

it('FetchApiForm should match snapshot', () => {
    const component = renderer.create(<FetchApiForm/>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

it('Button should have correct text', () => {
    const wrapper = mount(<FetchApiForm/>);
    expect(wrapper.find("button").text()).toEqual('What is jest?');
});

it('Button should call handler', () => {
    const wrapper = mount(<FetchApiForm/>);
    const instance = wrapper.instance();
    const spy = jest.spyOn(instance, 'getInformation');
    instance.forceUpdate();
    wrapper.find('button').simulate('click');
    expect(spy).toHaveBeenCalledTimes(1);
});

describe('FetchApiForm test suite', () => {

    it('Fetch must be successful', done => {
        const message = "Fetched successfully";
        const mockResponse = {
            info: message
        };

        const mockJsonPromise = Promise.resolve(mockResponse);
        const mockFetchPromise = Promise.resolve({
            json: () => mockJsonPromise,
        });
        global.fetch = jest.fn().mockImplementation(() => mockFetchPromise);

        const wrapper = mount(<FetchApiForm/>);
        wrapper.instance().fetchApi(URL);

        expect(global.fetch).toHaveBeenCalledTimes(1);
        expect(global.fetch).toHaveBeenCalledWith(URL);

        process.nextTick(() => {
            expect(wrapper.state()).toEqual(
                {
                    error: [],
                    response: message
                }
            );

            global.fetch.mockClear();
            delete global.fetch;
            done();
        });
    });

    it('Fetch should produce error', done => {
        const errorMessage = "we were unable to find any matching requests for this method type and the mock path, '/info_dummy', in your collection.";
        const mockResponse = {
            error: {
                message: errorMessage
            }
        };

        const mockJsonPromise = Promise.resolve(mockResponse);
        const mockFetchPromise = Promise.resolve({
            json: () => mockJsonPromise,
        });
        global.fetch = jest.fn().mockImplementation(() => mockFetchPromise);

        const wrapper = mount(<FetchApiForm/>);
        wrapper.instance().fetchApi(URL);

        expect(global.fetch).toHaveBeenCalledTimes(1);
        expect(global.fetch).toHaveBeenCalledWith(URL);

        process.nextTick(() => {
            expect(wrapper.state()).toEqual(
                {
                    error: errorMessage,
                    response: []
                }
            );

            global.fetch.mockClear();
            delete global.fetch;
            done();
        });
    });

    it('Fetch should catch an error', done => {
        const errorMessage = "we were unable to find any matching requests for this method type and the mock path, '/info_dummy', in your collection.";
        const mockResponse = {
            message: errorMessage
        };

        const mockFetchPromise = Promise.reject(mockResponse);
        global.fetch = jest.fn().mockImplementation(() => mockFetchPromise);

        const wrapper = mount(<FetchApiForm/>);
        wrapper.instance().fetchApi(URL);

        expect(global.fetch).toHaveBeenCalledTimes(1);
        expect(global.fetch).toHaveBeenCalledWith(URL);

        process.nextTick(() => {
            expect(wrapper.state()).toEqual(
                {
                    error: errorMessage,
                    response: []
                }
            );

            global.fetch.mockClear();
            delete global.fetch;
            done();
        });
    });
});