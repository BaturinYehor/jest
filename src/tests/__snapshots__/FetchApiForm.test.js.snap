// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FetchApiForm should match snapshot 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    What is jest?
  </button>
  <div
    className="get-api-form"
  >
    <div
      className="text"
    >
      <div
        className="error"
      />
    </div>
  </div>
</div>
`;
